src/transports/ws/sws.c

```diff
@@ -773,14 +744,21 @@ static int nn_sws_fail_conn (struct nn_sws *self, int code, char *reason)
 
     self->fail_msg_len += payload_len;
 
-    self->instate = NN_SWS_INSTATE_FAILING;
+    self->instate = NN_SWS_INSTATE_CLOSED;
 
-    /*  On the next recv, the connection will be failed. Why defer
-        until the next recv? Semantically, until then, this incoming
-        message has not been interpreted, so it's not until then that
-        it could be failed. This type of pre-processing is necessary
-        to early fail chunked transfers. */
```

Comment:
> got back a zero length frame which was NOT what should have happened

Aha. I think I hear where you're coming from now, and probably agree with what you did here.

As for the zero-length message -- the *payload* is zero length, but the message carried information in the CMSG header about the disconnected peer. Which was the application's way of detecting the peer's "close handshake", yet i think we've come to the agreement that the spirit of nanomsg should not reveal this to the application layer.